tests/unit/s2n_self_talk_ktls_test.c

```diff
@@ -0,0 +1,250 @@
+        EXPECT_SUCCESS(s2n_fd_set_non_blocking(io_pair.server));
+        EXPECT_SUCCESS(s2n_fd_set_non_blocking(io_pair.client));
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server, client));
+
+        if (s2n_connection_ktls_enable_send(client) != S2N_SUCCESS) {
+            /* Even if we detected ktls support at compile time, enabling ktls
+             * can fail at runtime if the system is not properly configured.
+             */
+            EXPECT_FALSE(ktls_expected);
+            END_TEST();
+        }
+        EXPECT_SUCCESS(s2n_connection_ktls_enable_send(server));
+    };
+
+    /* Test sending with ktls */
+    for (size_t mode_i = 0; mode_i < s2n_array_len(modes); mode_i++) {
+        const s2n_mode mode = modes[mode_i];
+
+        DEFER_CLEANUP(struct s2n_connection *client = s2n_connection_new(S2N_CLIENT),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client);
+        EXPECT_SUCCESS(s2n_connection_set_config(client, config));
+
+        DEFER_CLEANUP(struct s2n_connection *server = s2n_connection_new(S2N_SERVER),
+                s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client);
+        EXPECT_SUCCESS(s2n_connection_set_config(server, config));
+
+        DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+        EXPECT_OK(s2n_new_inet_socket_pair(&io_pair));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client, server, &io_pair));
+
+        /* The test negotiate method assumes non-blocking sockets */
+        EXPECT_SUCCESS(s2n_fd_set_non_blocking(io_pair.server));
+        EXPECT_SUCCESS(s2n_fd_set_non_blocking(io_pair.client));
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server, client));
+
+        struct s2n_connection *conns[] = {
+            [S2N_CLIENT] = client,
+            [S2N_SERVER] = server,
+        };
+        struct s2n_connection *writer = conns[mode];
+        struct s2n_connection *reader = conns[S2N_PEER_MODE(mode)];
+        EXPECT_SUCCESS(s2n_connection_ktls_enable_send(writer));
+
+        s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+
+        /* Our IO methods are more predictable if they use blocking sockets. */
+        EXPECT_SUCCESS(s2n_fd_set_blocking(io_pair.server));
+        EXPECT_SUCCESS(s2n_fd_set_blocking(io_pair.client));
+
+        /* Test: s2n_send */
+        for (size_t i = 0; i < 5; i++) {
+            int written = s2n_send(writer, test_data, sizeof(test_data), &blocked);
+            EXPECT_EQUAL(written, sizeof(test_data));
+            EXPECT_EQUAL(blocked, S2N_NOT_BLOCKED);
+
+            uint8_t buffer[sizeof(test_data)] = { 0 };
+            int read = s2n_recv(reader, buffer, sizeof(buffer), &blocked);
+            EXPECT_EQUAL(read, sizeof(test_data));
+            EXPECT_EQUAL(blocked, S2N_NOT_BLOCKED);
+
+            EXPECT_BYTEARRAY_EQUAL(test_data, buffer, read);
+        }
+
+        /* Test: s2n_sendv_with_offset */
```

Comment:
If the normal case for ktls is to call `s2n_sendv` directly, should there be an `s2n_sendv` test? Or is this not needed since `s2n_sendv` just calls `s2n_sendv_with_offset`?

Reply:
The way I was constructing iovecs, it would have basically been a duplicate of s2n_send. But I moved the test iovec builder I wrote for s2n_ktls_io_test to the testlib and used that to make more interesting s2n_sendv tests (and more interesting s2n_sendv_with_offset tests :P).